### PR TITLE
[RISCV] Optimize gather/scatter to unit-stride memop + shuffle

### PR DESCRIPTION
**Pull request by preames**
If we have a gather or a scatter whose index describes a permutation of the lanes, we can lower this as a shuffle + a unit strided memory operation.  For RISCV, this replaces a indexed load/store with a unit strided memory operation and a vrgather (at worst).

I did not both to implement the vp.scatter and vp.gather variants of these transforms because they'd only be legal when EVL was VLMAX.  Given that, they should have been transformed to the non-vp variants anyways.  I haven't checked to see if they actually are.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -13510,6 +13510,40 @@ static bool legalizeScatterGatherIndexType(SDLoc DL, SDValue &Index,
...
+  if (!ISD::isBuildVectorOfConstantSDNodes(Index.getNode()))
+    return false;
+
+  const unsigned ElementSize = VT.getScalarStoreSize();
+  const unsigned NumElems = VT.getVectorNumElements();
+
+  // Create the shuffle mask and check all bits active
+  assert(ShuffleMask.empty());
+  BitVector ActiveLanes(NumElems);
+  for (const auto Idx : enumerate(Index->ops())) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:13530]: Why are we using the ops iterator but then discarding the value() part? Can we just use `for (unsigned i = 0; i < Index->getNumOperands(); ++i)`?

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: I pushed a change to fix this.

**Review by topperc — APPROVED** [review]
LGTM

**Review by topperc — APPROVED** [review]
LGTM

**Comment by preames** [comment]
Pushed as ff2622b5ac6befc1f0814f83dabed8c04d7215ef.

**Pull request closed without merging**